docs/Architecture.md

```diff
@@ -0,0 +1,200 @@
+
+The actor model provides a clean separation between components, allowing eclair to:
+
+1. Isolate faults and ensure high availability
+2. Scale across CPUs and machines efficiently
+3. Simplify development and testing
+
+At a high-level, almost every entity is a separate, sandboxed actor:
+
+- Every peer connection is an actor instance
+- Every lightning channel is an actor instance
+- Every payment attempt is an actor instance
+
+Some actors are long-lived (e.g. lightning channels) while others are very short-lived (e.g. payment attempts).
+
+## Top-level projects
+
+Eclair is split into four top-level projects:
+
+- `eclair-core`: core library implementing lightning
+- `eclair-node`: server daemon built upon `eclair-core` (exposes a Json RPC and WebSocket endpoint)
+- `eclair-front`: when using cluster mode, front-end server daemons handling peer connections
+- `eclair-node-gui` (deprecated): sample JavaFX user interface to demo eclair (should not be used in production)
+
+The entry point for `eclair-core` is in `Setup.scala`, where we start the actor system, connect to `bitcoind` and create top-level actors.
+
+## Actor system overview
+
+Here is a high-level view of the main actors in the system:
+
+```ascii
+                                                             +--------+
+                          +--------------------------------->| Router |<------------------------------------+
+                          |                                  +--------+                                     |
+                          |                                                                                 |
+                          |           +---------+                                                           |
+                          | +-------->| Channel |<----------------+                                         |
+                          v |         +---------+                 |                                         |
+                        +------+      +---------+                 |                                         v
+      +---------------->| Peer |----->| Channel |<----------------+                                +------------------+
+      |                 +------+      +---------+                 |                    +-----------| PaymentInitiator |
+      |                     |         +---------+                 |                    |           +------------------+
+      |                     +-------->| Channel |<----------------+                    |
+      |                               +---------+                 |                    v
++-------------+                                              +----------+         +---------+
+| Switchboard |                                              | Register |<--------| Relayer |
++-------------+                                              +----------+         +---------+
```

Comment:
Channels also directly send messages to the `Relayer`. On the other hand, the `PaymentInitiator` sends messages to the `Register` (but receive messages from the  `Relayer`.

Reading further I noticed that you detail the send/receive/relay scenarii below. Maybe in this diagram we should represent the actor hierarchy (it is a bit unclear what the arrows represent).

I wonder if it would be useful to also mention the `PeerConnection`/`TransportHandler`.

Reply:
I tried to keep this high-level, so that a newcomer can easily locate an entry point in the code based on what he's looking for. That's why I didn't mention `PeerConnection`/`TransportHandler`, if you're looking for them you'll figure out from the diagram that you should look into the `Switchboard` or the `Peer` and then the code will lead you to these lower-level actors.

I tried to show a simple message flow, but as you noticed it's actually incorrect and when you dive into the details it's always a bit more complex than we'd like...I'll try changing this diagram to expose the actor hierarchy as you suggest, it's probably easier.

Reply:
Done in https://github.com/ACINQ/eclair/pull/1733/commits/5542fdf84852e89869db6f97759acb11e629d32d, it's a bit hard to decide where to stop to keep this high-level but detailed enough to give a good overall view...